src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections;
+using System;
+using System.Diagnostics;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public partial interface IOperation
+    {
+        /// <summary>
+        /// Implements a struct-based collection of <see cref="Operation"/> nodes. This collection is ordered, but
+        /// random access into the collection is not provided. This type is not hardened to 
+        /// <code>default(ChildCollection)</code>, and will null reference in these cases.
+        /// </summary>
+        [NonDefaultable]
+        public readonly partial struct ChildCollection : IReadOnlyCollection<IOperation>
+        {
+            private readonly Operation _operation;
+
+            internal ChildCollection(Operation operation)
+            {
+                _operation = operation;
+            }
+
+            public int Count => _operation.ChildOperationsCount;
+
+            public Enumerator GetEnumerator() => new Enumerator(_operation);
+
+            public ImmutableArray<IOperation> ToImmutableArray()
+            {
+                Enumerator enumerator = GetEnumerator();
+                switch (_operation)
+                {
+                    case NoneOperation { Children: var children }:
+                        return children;
+                    case InvalidOperation { Children: var children }:
+                        return children;
+                    case var _ when !enumerator.MoveNext():
```

Comment:
>when !enumerator.MoveNext() [](http://example.com/codeflow?start=31&length=27)

It looks like the `default` branch depends on the execution order of the `case` and `when` clauses, which seems undefined. Please consider checking `Count == 0` explicitly instead, and move the `Enumerator` instance into the `default` branch.

Reply:
Could you please clarify this? It seems like since `var _` always matches, it is well-defined that the 'default' label here is only reachable if `!enumerator.MoveNext()` is evaluated and returned `false`.

If you object as a matter of style, that's understandable, but I personally don't have an issue with it.

Reply:
Does the compiler guarantee that a `when` clause is executed at most once? And is it good practice to have a `when` clause that has side-effects, such as modifying `enumerator`?

We should avoid these questions entirely, and simply refactor the code. For instance, either use `when Count == 0:` here or combine this `case` and the `default`:
```csharp
default:
    if (Count == 0)
        return ImmutableArray<IOperation>.Empty;
    var builder = ArrayBuilder<IOperation>.GetInstance(Count);
    ...
    return builder.ToImmutableAndFree();
```

Reply:
> Does the compiler guarantee that a when clause is executed at most once?

That is how the implementation behaves, and I would be very surprised if we ever changed that intentionally. The spec is very light on this, which would be good to address in another venue. There is a considerable compatibility burden for what we have shipped regardless of what we have declared to be implementation-defined.

I just don't want us to treat these constructs as something which may start behaving in an incompatible manner in next year's compiler. At this point any breaking change that we make to patterns would have to be something people are very unlikely to be depending on.

> We should avoid these questions entirely, and simply refactor the code.

I would like for us to have an "official" answer for the above question in a separate venue from this PR. As far as our actual development practices go, what you've suggested works for me. I can't argue against the claim that `enumerator.MoveNext()` in a when clause is harder to follow.